content/shmem_init_thread.tex

```diff
@@ -0,0 +1,78 @@
+\apisummary{
+Initializes the \openshmem{} library, similar to SHMEM\_INIT(), and in addition
+perform the initialization required for thread-safe invocation of \openshmem{} functions.}
+
+\begin{apidefinition}
+
+\begin{Csynopsis}
+int shmem_init_thread(int requested, int *provided);
+\end{Csynopsis}
+
+\begin{Fsynopsis}  
+INTEGER requested, provided
+     
+SHMEM_INIT_THREAD(requested, provided)
+\end{Fsynopsis}  
+
+\begin{apiarguments}
+\apiargument{IN}{requested}  {The thread level support requested by the user.
+The correct values are SHMEM\_THREAD\_SINGLE, SHMEM\_THREAD\_FUNNELED, or SHMEM\_THREAD\_MULTIPLE}
+\apiargument{OUT}{provided}{The thread level support provided by the \openshmem{} implementation.}
+\end{apiarguments}
+
+
+\apidescription{
+    \FUNC{shmem\_init\_thread}  initializes the \openshmem{} library similar to
+    \FUNC{shmem\_init}, and in addition perform the initialization required for
+    thread-safe invocation of \openshmem{} functions. The argument
+    \textit{requested} is used to specify the desired level of thread support.
+    The function returns the support provided by the library. There are four 
+    levels of threading support.
+ 
+ \begin{itemize}
+\item SHMEM\_THREAD\_SINGLE: The \openshmem{} program may not be multithreaded, 
+and all \openshmem{} interfaces are invoked by a single thread. 
+
+\item SHMEM\_THREAD\_FUNNELED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces.
+
+\item SHMEM\_THREAD\_SERIALIZED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces at any given instance i.e., the \openshmem{} interfaces 
+are not invoked concurrently by multiple threads.
+
+\item SHMEM\_THREAD\_MULTIPLE: 
+The \openshmem{} program may be multithreaded, and any 
+thread may invoke the \openshmem{} interfaces.
+
+\item SHMEM\_THREAD\_UNSUPPORTED: The \openshmem{} specification 
```

Comment:
I agree that it's different from SHMEM_THREAD_SINGLE, but it's also not informative, which I think is my fundamental issue. If a developer is calling `shmem_init_thread()` or `shmem_query_thread()`, then they're already using an API beyond 1.3.  A return value of SHMEM_THREAD_UNSUPPORTED does not tell the developer anything meaningful about whether any threads can be used and, if so, whether those threads can make OpenSHMEM API calls.

An OpenSHMEM implementation that wants to make no substantial effort for thread support should _at a minimum_ identify which of the four models (likely, SINGLE or FUNNELED) is the default level of support.